phoenix-pherf/src/main/java/org/apache/phoenix/pherf/workload/mt/tenantoperation/TenantOperationFactory.java

```diff
@@ -0,0 +1,501 @@
+
+                        } catch (Throwable t) {
+                            // swallow;
+                        }
+                    }
+                    return new OperationStats(input, startTime, 0, resultRowCount, queryElapsedTime);
+                }
+            };
+        }
+    }
+
+    class UpsertTenantOperationImpl implements TenantOperationImpl {
+
+        @Override public Function<TenantOperationInfo, OperationStats> getMethod() {
+            return new Function<TenantOperationInfo, OperationStats>() {
+
+                @Nullable @Override public OperationStats apply(@Nullable TenantOperationInfo input) {
+
+                    final int batchSize = loadProfile.getBatchSize();
+                    final boolean useBatchApi = batchSize != 0;
+                    final int rowCount = useBatchApi ? batchSize : 1;
+
+                    final UpsertOperation operation = (UpsertOperation) input.getOperation();
+                    final String tenantGroup = input.getTenantGroupId();
+                    final String opGroup = input.getOperationGroupId();
+                    final String tenantId = input.getTenantId();
+                    final Upsert upsert = operation.getUpsert();
+                    final String tableName = input.getTableName();
+                    final String scenarioName = input.getScenarioName();
+                    final List<Column> columns = upsert.getColumn();
+
+                    final String opName = String.format("%s:%s:%s:%s:%s",
+                            scenarioName, tableName, opGroup, tenantGroup, tenantId);
+
+                    long rowsCreated = 0;
+                    long startTime = 0, duration, totalDuration;
+                    SimpleDateFormat simpleDateFormat = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
+                    try (Connection connection = phoenixUtil.getConnection(tenantId)) {
+                        connection.setAutoCommit(true);
+                        startTime = EnvironmentEdgeManager.currentTimeMillis();
+                        String sql = phoenixUtil.buildSql(columns, tableName);
+                        PreparedStatement stmt = null;
+                        try {
+                            stmt = connection.prepareStatement(sql);
+                            for (long i = rowCount; i > 0; i--) {
+                                LOGGER.debug("Operation " + opName + " executing ");
+                                stmt = phoenixUtil.buildStatement(rulesApplier, scenario, columns, stmt, simpleDateFormat);
+                                if (useBatchApi) {
+                                    stmt.addBatch();
+                                } else {
+                                    rowsCreated += stmt.executeUpdate();
+                                }
+                            }
+                        } catch (SQLException e) {
+                            LOGGER.error("Operation " + opName + " failed with exception ", e);
```

Comment:
Might be better to just let the outer catch block catch the exception since you're rethrowing it anyways. And then we can log there.